include/revng/RestructureCFG/ScopeGraphUtils.h

```diff
@@ -0,0 +1,190 @@
+    llvm::LLVMContext &C = getContext(F);
+    llvm::Module *M = getModule(F);
+    FunctionTags::Tag &Tag = FunctionTags::ScopeCloserMarker;
+    ScopeCloserFunction = getUniqueFunctionWithTag(Tag, M);
+
+    // Create the `ScopeCloserMarker` function if it doesn't exists
+    if (not ScopeCloserFunction) {
+      auto *FT = llvm::FunctionType::get(llvm::Type::getVoidTy(C), {}, false);
+      using llvm::Function;
+      using llvm::GlobalValue;
+      ScopeCloserFunction = llvm::cast<
+        llvm::Function>(M->getOrInsertFunction(Tag.name(), FT).getCallee());
+      setFunctionAttributes(ScopeCloserFunction, Tag);
+    }
+  }
+
+  void initGotoBlockFunction(llvm::Function *F) {
+    llvm::LLVMContext &C = getContext(F);
+    llvm::Module *M = getModule(F);
+    FunctionTags::Tag &Tag = FunctionTags::GotoBlockMarker;
+
+    GotoBlockFunction = getUniqueFunctionWithTag(Tag, M);
+
+    // Create the `ScopeCloserMarker` function if it doesn't exists
+    if (not GotoBlockFunction) {
+      llvm::PointerType *BlockAddressTy = llvm::Type::getInt8PtrTy(C);
+      auto *FT = llvm::FunctionType::get(llvm::Type::getVoidTy(C),
+                                         { BlockAddressTy },
+                                         false);
+      using llvm::Function;
+      using llvm::GlobalValue;
+      GotoBlockFunction = llvm::cast<
+        llvm::Function>(M->getOrInsertFunction(Tag.name(), FT).getCallee());
+      setFunctionAttributes(GotoBlockFunction, Tag);
+    }
+  }
+
+public:
+  ScopeGraphBuilder(llvm::Function *F) {
+    initScopeCloserFunction(F);
+    initGotoBlockFunction(F);
+  }
+
+public:
+  void makeGoto(llvm::BasicBlock *GotoBlock) {
+    // We must have a `GotoBlock`
+    revng_assert(GotoBlock);
+
+    // We assume that when inserting a `goto` edge, the original block had a
+    // single regular successor on the CFG
+    llvm::Instruction *Terminator = GotoBlock->getTerminator();
+    revng_assert(Terminator->getNumSuccessors() == 1);
+
+    // We always insert the marker as the penultimate instruction in a
+    // `BasicBlock`
+    llvm::IRBuilder<> Builder(Terminator);
+    llvm::LLVMContext &C = getContext(GotoBlock);
+    llvm::PointerType *BlockAddressTy = llvm::Type::getInt8PtrTy(C);
+    auto
+      *NullBasicBlockAddress = llvm::ConstantPointerNull::get(BlockAddressTy);
+    Builder.CreateCall(GotoBlockFunction, NullBasicBlockAddress);
+  }
+
+  void addScopeCloser(llvm::BasicBlock *Source, llvm::BasicBlock *Target) {
+    // We must have an insertion point
+    revng_assert(Source);
+
+    // We always insert the marker as the penultimate instruction in a
+    // `BasicBlock`
+    llvm::Instruction *Terminator = Source->getTerminator();
+    llvm::IRBuilder<> Builder(Terminator);
+    auto *BasicBlockAddressTarget = llvm::BlockAddress::get(Target);
+    revng_assert(BasicBlockAddressTarget);
+    Builder.CreateCall(ScopeCloserFunction, BasicBlockAddressTarget);
+  }
+};
```

Comment:
I moved all was possible to the `.cpp` file, except the templated helper which I now constrained with the `ConstOrNot` concept https://github.com/revng/revng/blob/fe8a4d7ebc61d0ceb8d64cdfe5c3ecf8fd1c21f4/include/revng/RestructureCFG/ScopeGraphUtils.h#L19.

Just a note regarding the 3 private methods, just ``setFunctionAttributes` can become a `static` free function, because `initScopeCloserFunction` and `initGotoBlockFunction` are needed to initialize the `ScopeCloserFunction` and `GotoBlockFunction` private fields of the `ScopeGraphBuilder`, which are needed when adding a marker.

Reply:
There still room for improvement. See the last comment I just opened.

Reply:
Continued this discussion on the other thread